Add a new stable chart "weave-scope".

omkensey opened the pull request:
This PR adds a `weave-scope` chart with two subcharts (`weave-scope-frontend` and `weave-scope-agent`).

Commit:
Add a new stable chart "weave-scope".

Commit:
Add a new stable chart "weave-scope".

omkensey commented:
(This is a resubmission of a previous PR, #1625 with the edits suggested there incorporated into this PR.)

unguiculus commented:
/assign

unguiculus commented on stable/weave-scope/Chart.yaml:
```diff
@@ -0,0 +1,15 @@
+sources:
+  - https://github.com/weaveworks/scope
+maintainers:
+  - name: Joe Thompson
```
Please use Github username.

unguiculus commented on stable/weave-scope/NOTES.txt:
```diff
@@ -0,0 +1,19 @@
+After installing a Scope frontend and node agents, access the frontend 
+in your web browser.  If you used this chart to install the frontend, 
+it can be accessed in the following ways:
+
```
Can you make the output dependent on how the user installed the chart and print the necessary commands so the user can cut and paste them?

omkensey replied:
Ah right, I forgot NOTES.txt gets rendered too!  Yes, will do.

unguiculus commented on stable/weave-scope/README.md:
```diff
@@ -0,0 +1,70 @@
+## Values templated in this chart
+
+Note that most of this documentation is repeated in `values.yaml`; if you're in a hurry you can skip this part here and read it there.  Values with no default noted have no default.
+
```
Why not use a table as it is commonly done?

omkensey replied:
Just didn't think of it.  Will do this too.

unguiculus commented on stable/weave-scope/charts/weave-scope-agent/templates/daemonset.yaml:
```diff
@@ -0,0 +1,72 @@
+    matchLabels:
+      name: {{ template "fullname" . }}-agent
+      app: {{ template "fullname" . }}
+      heritage: {{ .Release.Service }}
```
Remove `heritage` and `name` from selector. Add `release`.

unguiculus commented on stable/weave-scope/templates/_helpers.tpl:
```diff
@@ -0,0 +1,53 @@
+chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+heritage: {{ .Release.Service }}
+release: {{ .Release.Name }}
+app: {{ template "fullname" . }}
```
`app: {{ template "name" . }}`

unguiculus commented on stable/weave-scope/charts/weave-scope-agent/templates/daemonset.yaml:
```diff
@@ -0,0 +1,72 @@
+  selector: 
+    matchLabels:
+      name: {{ template "fullname" . }}-agent
+      app: {{ template "fullname" . }}
```
`app: {{ template "name" . }}`

omkensey replied:
I don't have a template called "name" I don't think, I'll add that one in from the nginx example and make this and the other corresponding changes.

unguiculus commented on stable/weave-scope/charts/weave-scope-agent/templates/daemonset.yaml:
```diff
@@ -0,0 +1,72 @@
+      labels:
+        name: {{ template "fullname" . }}-agent
+        {{- include "helm_std_labels" . | indent 8 }}
+        component: frontend
```
`component: agent`?

omkensey replied:
Yep, I forgot to change that one when I c&p'ed.  Will fix.

unguiculus commented on stable/weave-scope/charts/weave-scope-agent/templates/daemonset.yaml:
```diff
@@ -0,0 +1,72 @@
+    metadata:
+      labels:
+        name: {{ template "fullname" . }}-agent
+        {{- include "helm_std_labels" . | indent 8 }}
```
Only `app`, `release`, and `component` should go here.

unguiculus reviewed (changes requested):
Please apply my comments on the daemonset to other resources as well. Why do you add `name` labels? I don't like that. It is redundant.

omkensey commented:
The `name:` labels are actually in the original Weave Scope Kubernetes manifest published by Weaveworks that this is derived from.  I assume that in that case they're there to allow querying for pods by name, but with the Helm-related labels that's probably not needed; I don't have a strong attachment to them and if you don't like them I will certainly remove them.

unguiculus commented:
I'd rather remove them to be honest.

Commit:
Updates requested by @unguiculus, plus various bug fixes.

I believe this fixes all the issues raised against the initial PR.  Other changes of note:

* Child charts now have aliases so they will not have hyphens when used as values.
* Removed the clusterDomain value since it's not actually needed for the intended use.
* Version bump to 0.9.0.

Commit:
Updates requested by @unguiculus, plus various bug fixes.

I believe this fixes all the issues raised against the initial PR.  Other changes of note:

* Child charts now have aliases so they will not have hyphens when used as values.
* Removed the clusterDomain value since it's not actually needed for the intended use.
* Version bump to 0.9.0.

unguiculus commented on stable/weave-scope/charts/weave-scope-frontend/templates/deployment.yaml:
```diff
@@ -0,0 +1,32 @@
+          ports:
+            - containerPort: {{ .Values.containerPort | default 4040 }}
+              protocol: TCP
+          {{- include "scope_resources" . | indent 10 }}
```
There's no need for a template function. That would just make it overly complicated. This is how it is commonly done:

```yaml
resources:
{{ toYaml .Values.resources | indent 12 }}
```

unguiculus reviewed (changes requested):
You specify default values in lots of places using the `default` template function. It would be better to add the defaults to the `values.yaml` instead.

omkensey commented:
Any issue with having the defaults be specified in the templates and also having them in `values.yaml`?  I want it to do something reasonable even if the user passes no values at all.

unguiculus commented:
We tend to recommend not to use the `default` function if possible and to only have the defaults in `values.yaml`. Doing both would be redundant.

omkensey commented:
OK, I got that from the line in the chart template guide that says "It is considered good (almost mandatory) practice to set defaults with default for any object that originates from .Values."  Can change it if you prefer though.

unguiculus commented:
That's exactly what I am suggesting. Use defaults from `.Values`. Using the `default` function on top of that is unnecessary and redundant.

scottrigby commented:
I agree with @unguiculus on this. But I can see why @omkensey is saying
> But… the docs…

😄 

I opened an issue+PR to help resolve this confusion: https://github.com/kubernetes/helm/pull/2889

unguiculus commented:
@omkensey I misread, our comment.

Thanks, @scottrigby, your PR makes it clear.

omkensey commented:
Given that, is it still considered necessary or good practice to wrap calls to values nested beyond the first level, that have defaults in values.yaml, in `{{- if }}` blocks like so?

```
{{- if .Values.global.service }}
    - name: {{ .Values.global.service.name }}
      port: {{ .Values.global.service.port }}
{{- else }}
    - name: "weave-scope"
      port: 80
{{- end }}
```

...or can I just drop the conditionals entirely?

scottrigby commented:
@omkensey Cross-posting my reply to your cross-post. See https://github.com/kubernetes/helm/issues/2888#issuecomment-326820735

unguiculus commented:
Yes, please drop the conditionals whenever possible and add defaults only to the `values.yaml`.

gyliu513 commented:
This fixed https://github.com/kubernetes/charts/issues/2118

unguiculus commented:
Marking this as stale. please update within one week.

omkensey commented:
Working on update, but having some issues with the way the alias feature works.  Will update as soon as I can get it to render...

unguiculus commented:
@omkensey Any progress? Otherwise I should really close the PR and you may reopen it once you make it work.

omkensey commented:
Yes, sorry, I have a workaround that looks good and I'll be working on this tonight and tomorrow to update :)

etiennetremel commented:
What's the status of this PR? I would love to use that setup :)

omkensey commented:
I have an update to check in tonight (which is to say early AM UTC) that will include tests and a little refactoring.  Sorry for the long delay!

Commit:
Updates and workarounds, plus added a test.

Commit:
Updates and workarounds, plus added a test.

unguiculus commented on stable/weave-scope/README.md:
```diff
@@ -0,0 +1,65 @@
+
+The **weave-scope-frontend** section controls how the Scope frontend is installed.
+
+| Value | Default | Explanation |
```
Please go with the style that is commonly used, which is 
`| Parameter | Description | Default |`.

unguiculus commented on stable/weave-scope/templates/test-config.yaml:
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "weave-scope.fullname" . }}-tests
```
Add standard labels

unguiculus commented on stable/weave-scope/templates/weave-scope-tests.yaml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: "{{ .Release.Name }}-ui-test-{{ randAlphaNum 5 | lower }}"
```
Add standard labels

unguiculus commented on stable/weave-scope/requirements.yaml:
```diff
@@ -0,0 +1,10 @@
+# We alias the child chart names to avoid having to deal with hyphenated 
+# YAML keys.
+
```
This file is unnecessary. It is used for dynamically linked dependencies only, whereas you added the dependencies directly to the charts folder. I guess that's the reason why you can't use aliases in this case.

omkensey replied:
Maybe that's it.  There's some discussion related to aliases going on in kubernetes/helm#2993.  IMHO it should be possible to alias directly-added subcharts, but for now since I'm not actually using the aliases I can remove this.

omkensey replied:
Actually I just realized there is a reason to still have this file -- I'm using conditionals to allow for disabling deployment of either of the subcharts.  I deleted the "alias" lines though.

unguiculus reviewed (changes requested):
What's the reason for modeling this with dependencies? I don't see any advantage since they are directly added instead of dynamically linked. It seems it just makes things more complicated.

Also, please follow these examples for RBAC:
https://github.com/coreos/alb-ingress-controller/pull/146
https://github.com/kubernetes/charts/pull/1504

Commit:
* Added standard Helm labels to test resources
* Removed aliasing from `requirements.yaml`
* Reformatted tables in `README.md`
* Updates throughout:
  * Current version of Weave Scope (1.6.5)
  * Current version of upstream manifest (added a debug volume/mount to the agent)

Commit:
* Added standard Helm labels to test resources
* Removed aliasing from `requirements.yaml`
* Reformatted tables in `README.md`
* Updates throughout:
  * Current version of Weave Scope (1.6.5)
  * Current version of upstream manifest (added a debug volume/mount to the agent)

omkensey commented:
I'll commit the RBAC changes later tonight.

Commit:
Redid RBAC; fixed containerPort bug.

Commit:
Redid RBAC; fixed containerPort bug.

Commit:
Merge branch 'weave-scope' of github.com:omkensey/charts into weave-scope

Commit:
Fix trailing space error.

unguiculus committed:
Merge remote-tracking branch 'upstream/master' into HEAD

unguiculus commented on stable/weave-scope/values.yaml:
```diff
@@ -0,0 +1,76 @@
+# Where defaults exist, the values are set to them here.
+# Values with no preferred or common defaults are set to empty strings.
+global:
+# global.image: (required) the image that will be used for this release
```
Please indent comments as well to make things clearer.

omkensey replied:
What's the preferred format?

```
# foo: comment
foo: 1
bar:
  # bar.baz: comment
  baz: 2
```

or something else?

unguiculus commented on stable/weave-scope/values.yaml:
```diff
@@ -0,0 +1,76 @@
+# weave-scope-agent.rbac.create: whether RBAC resources should be created
+# weave-scope-agent.rbac.create *must* be set to false if RBAC is not enabled in the cluster
+# weave-scope-agent.rbac.create *may* be set to false in an RBAC-enabled cluster to allow for external management of RBAC
+    create: true
```
Default should be `false`.

omkensey replied:
RBAC is enabled by kubeadm by default starting with 1.6 (other cluster installers like Tectonic enabled it earlier than that, even) so I figured with that change the default for rbac.create should match the default for a cluster -- but if you feel strongly about it, I'll change it.  (One of the example chart PRs you linked to above has it false, one has it true, so I don't know if there's a true convention around it yet.)

unguiculus replied:
Hm, I checked several charts and only found it to be `false` by default. I don't have a strong opinion on this.

omkensey replied:
I filed kubernetes/helm#3098 on this.

unguiculus commented on stable/weave-scope/charts/weave-scope-frontend/templates/deployment.yaml:
```diff
@@ -0,0 +1,40 @@
+      containers:
+        - name: app
+          image: {{ required "You must specify the desired container image" .Values.global.image }}:{{ required "You must specify the desired container image tag" .Values.global.imageTag }}
+          {{- if .Values.imagePullPolicy }}
```
Remove the if statement and set `IfNotPresent` in `values.yaml`.

unguiculus commented on stable/weave-scope/charts/weave-scope-frontend/templates/deployment.yaml:
```diff
@@ -0,0 +1,40 @@
+    spec:
+      containers:
+        - name: app
+          image: {{ required "You must specify the desired container image" .Values.global.image }}:{{ required "You must specify the desired container image tag" .Values.global.imageTag }}
```
Image and tag are already specified in `values.yaml`. I would remove the `required` calls.

unguiculus commented on stable/weave-scope/charts/weave-scope-agent/templates/daemonset.yaml:
```diff
@@ -0,0 +1,86 @@
+          operator: Exists
+      containers:
+        - name: agent
+          image: {{ required "You must specify the desired container image" .Values.global.image }}:{{ required "You must specify the desired container image tag" .Values.global.imageTag }}
```
Image and tag are already specified in `values.yaml`. I would remove the `required` calls.

unguiculus commented on stable/weave-scope/charts/weave-scope-agent/templates/daemonset.yaml:
```diff
@@ -0,0 +1,86 @@
+      containers:
+        - name: agent
+          image: {{ required "You must specify the desired container image" .Values.global.image }}:{{ required "You must specify the desired container image tag" .Values.global.imageTag }}
+          {{- if .Values.imagePullPolicy }}
```
Remove the if statement and set `IfNotPresent` in `values.yaml`.

omkensey replied:
Doc change needed?  [Chart Best Practices](https://docs.helm.sh/chart_best_practices/#imagepullpolicy) says it should default to an empty value.

unguiculus replied:
Interesting. I wasn't aware of that. Chart examples and the chart created by `helm create` use `IfNotPresent`.

omkensey replied:
I filed kubernetes/helm#3099 on this.

unguiculus commented on stable/weave-scope/charts/weave-scope-agent/templates/clusterrole.yaml:
```diff
@@ -0,0 +1,24 @@
+{{- if .Values.rbac }}
```
The outer if statement is unnecessary.

omkensey replied:
I have a similar structure in the daemonset template:

```
      {{- if .Values.rbac }}
        {{- if .Values.rbac.create }}
      serviceAccountName: {{ required "rbac.serviceAccountName must be set if rbac.create is true" .Values.rbac.serviceAccountName }}
        {{- else if .Values.rbac.serviceAccountName }}
      serviceAccountName: {{ .Values.rbac.serviceAccountName }}
        {{- end }}
      {{- end }}
```

Since the only difference in the behavior of the template between `rbac.create: true` and `rbac.create: false` is putting a "required" on `rbac.serviceAccountName`, and there's already a value for that set in values.yaml, can I just collapse that whole block back down to `serviceAccountName: {{ .Values.rbac.serviceAccountName }}` then?

unguiculus replied:
yes

unguiculus commented on stable/weave-scope/charts/weave-scope-agent/templates/clusterrolebinding.yaml:
```diff
@@ -0,0 +1,21 @@
+{{- if .Values.rbac }}
```
The outer if statement is unnecessary.

unguiculus commented on stable/weave-scope/charts/weave-scope-agent/templates/serviceaccount.yaml:
```diff
@@ -0,0 +1,13 @@
+{{- if .Values.rbac }}
```
The outer if statement is unnecessary.

unguiculus reviewed (changes requested):
Let me ask again why you model the chart with dependencies. This doesn't make much sense to me since they are directly added and not dynamically linked. I think this just makes things more complicated.

omkensey commented:
> Let me ask again why you model the chart with dependencies.

I'm using conditionals to allow for disabling deployment of either of the subcharts.  The only other way I know to do that is to use `{{- if }}` blocks on every template.

unguiculus commented:
Yes, using `{{- if }}` blocks would be the alternative. I'd prefer that in this case. The question is: When/why would it make sense to only install frontend or agent but not both? And if that makes sense, wouldn't it be better to create separate charts for them? And then potentially a third umbrella chart?

omkensey commented:
> When/why would it make sense to only install frontend or agent but not both?

Weave Scope can report to a cloud service rather than a local frontend (that's what the `probeToken` value in the weave-scope-agent subchart enables).  In that case you want just the agent, not the frontend.  There can also be cases where somebody has an existing local frontend and is deploying Scope to a new set of hosts and wants that set to report to the frontend they already have running.

On the other side, you might want to deploy just the frontend if you want to be able to update the frontend separately from the agents -- then you want the frontend in one release and the agents in their own release(s).

Using `weave-scope-{agent,frontend}.enabled` in requirements.yaml was just less writing to do (it will also be easier to update if the convention around conditional deployment of subcharts ever changes).

> And if that makes sense, wouldn't it be better to create separate charts for them? And then potentially a third umbrella chart?

It's tricky to do that because if you want to install both (and I think that *will* be the default case), then you have to make sure the agent chart can get the name of the service created by the frontend chart.  With them put together the way I have them, you can handle that with the `global.*` values set by the parent.  If the subcharts are written separately, then I have to have globals in each chart that don't really make sense there if they're deployed on their own.

Having said which, I came up with a hack that might actually do the job, but it requires messing with the standard format of the fullname helper template (or creating a new helper template) to read the actual file path of the chart where that template is set.  Seems kinda ugly.

unguiculus commented:
I'll try and experiment with it.

omkensey commented:
For the moment while discussion happens elsewhere, I'm going to make the changes I thumbed-up, plus setting `imagePullPolicy` to `IfNotPresent` and `rbac.create` to `false`.  If either of those needs changing after this PR is merged I'll do a new PR for it.

unguiculus commented:
What's the status now?

omkensey commented:
I'm finishing up a new commit.  It should be ready to push sometime tomorrow (18 Nov).

omkensey commented:
I've come up with this as a workaround for some of the naming issues I've run into:

```
{{- define "toplevel.fullname" -}}
{{- $name := default (.Template.BasePath | split "/" )._0 .Values.nameOverride -}}
{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" -}}
{{- end -}}
```

This will allow a "top-level" name referencing the topmost chart to be calculated the same in all subcharts.  It's a little hacky (see kubernetes/helm#3166) but it seems workable given the rules around chart directory naming.

Finishing up latest commit this afternoon.

Commit:
Changes requested by reviewer

Notable changes:
* Added "toplevel" templates used for consistency of certain generated names
* RBAC- and image-related values are structured differently
* Service name and service account name are no longer required
* containerPort is no longer configurable
* Superfluous conditionals and calls to `required` removed
* Ports are named
* Whitespace cleanup throughout

Commit:
Merge branch 'weave-scope' of github.com:omkensey/charts into weave-scope

Commit:
Whitespace fix

Trailing whitespace in values.yaml.

unguiculus commented:
/ok-to-test

omkensey commented:
Looks like I'm hitting, more or less, kubernetes/helm#2887.  I'll push a commit shortly that removes `requirements.yaml` and moves to `{{- if foo.enabled }}` blocks instead.

Commit:
Remove requirements.yaml

Refactor to use {{ if }} blocks instead of requirements.yaml to control child chart deployment.

unguiculus commented:
Does it really make sense to keep them as subcharts now?

omkensey commented:
I think so.  It doesn’t hurt usability, it’s easy for somebody to pull apart into components (e.g. I just used it that way to construct several demos for my talk at KubeCon) or fork into full independent charts, and if it becomes possible in the future to use requirements.yaml as I was trying to do originally, that will be more elegant than the current way.

unguiculus commented on stable/weave-scope/charts/weave-scope-frontend/templates/_helpers.tpl:
```diff
@@ -0,0 +1,50 @@
+chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+heritage: {{ .Release.Service }}
+release: {{ .Release.Name }}
+app: {{ template "toplevel.name" . }}
```
I don't like `toplevel`. I think it is too unspecific. Convention is to use the chart name, which would be `weave-scope`.

omkensey replied:
The trouble with using the conventional "chart-name.{name,fullname}" templates here is that even though the parent sets the definition all children use, they render differently in different children because .Chart.Name is based on the chart that renders the template, not the chart that defines it.  So the "app" attribute would be set differently per child chart, which is not what I want if the full Weave Scope chart is being used.  What I needed here was a reference to the parent chart name, that all children would render identically.

The idea behind the "toplevel" construct (cf. kubernetes/helm#3166) is to allow a chart to say "whatever top-level ancestor (parent, grandparent, etc.) chart is ultimately depending on us, use that chart's name here, or use the name of this chart if there is no ancestor" which is exactly what I was trying to do here (it's intentionally more generic than referring to the ancestor chart by name since a child won't necessarily know what ancestors it has).  If 3166 actually gets written and merged then there would be no use for the "toplevel" named templates as such, but in cases like this it provides a workaround where people have otherwise been asking how to export values from parent charts down to child charts (which Helm doesn't support yet as far as I know) so they can set something consistently in all children without needing to set globals.

unguiculus commented:
/retest

unguiculus commented:
I'm going to merge this now. Sorry it took so long.

unguiculus commented:
/lgtm
